src/App.jsx

```diff
@@ -1,36 +1,195 @@
+          }
+          return todo;
+        });
+        setTodoList([...fetchedTodos]);
+      } catch (error) {
+        setErrorMessage(error.message);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+    fetchTodos();
+  }, []);
+
+  async function handleAddTodo(title) {
     const newTodo = { title: title, id: Date.now(), isCompleted: false };
     setTodoList([...todoList, newTodo]);
+    const payload = {
+      records: [
+        {
+          fields: {
+            title: newTodo.title,
+            isCompleted: newTodo.isCompleted,
+          },
+        },
+      ],
+    };
+    const options = {
+      method: "POST",
+      headers: { Authorization: token, "Content-Type": "application/json" },
+      body: JSON.stringify(payload),
+    };
+    try {
+      setIsSaving(true);
+      const resp = await fetch(url, options);
+      if (!resp.ok) {
+        throw new Error(resp.message);
+      }
+      const { records } = await resp.json();
+      const savedTodo = { id: records[0].id, ...records.fields };
+      if (!records[0].fields.isCompleted) {
+        savedTodo.isCompleted = false;
+      }
+      setTodoList([...todoList, savedTodo]);
+    } catch (error) {
+      console.log(error);
+      setErrorMessage(error.message);
+    } finally {
+      setIsSaving(false);
+    }
   }
-  function completeTodo(id) {
-    const updatedTodos = todoList.map((todo) =>
-      todo.id === id ? { ...todo, isCompleted: true } : todo
-    );
-    setTodoList(updatedTodos);
+
+  async function handleUpdateTodo(editedTodo) {
+    const originalTodo = todoList.find((todo) => todo.id === editedTodo);
+    const payload = {
+      records: [
+        {
+          id: editedTodo.id,
+          fields: {
+            title: editedTodo.title,
+            isCompleted: editedTodo.isCompleted,
+          },
+        },
+      ],
+    };
+    const options = {
+      method: "PATCH",
+      headers: { Authorization: token, "Content-Type": "application/json" },
+      body: JSON.stringify(payload),
+    };
+    try {
+      const resp = await fetch(url, options);
+      if (!resp.ok) {
+        throw new Error(resp.status);
+      }
+      const { records } = await resp.json();
+      const updatedTodo = { id: records[0]["id"], ...records.fields };
+      if (!records[0].fields.isCompleted) {
+        updatedTodo.isCompleted = false;
+      }
+      const updatedTodos = todoList.map((todo) =>
+        todo.id === updatedTodo.id ? { ...updatedTodo } : todo
+      );
+      setTodoList([...updatedTodos]);
+    } catch (error) {
+      setErrorMessage(`${error.message}.Reverting todo...`);
+      const revertedTodos = setTodoList(
+        updatedTodos.map((todo) =>
+          todo.id === id ? { ...originalTodo } : todo
+        )
+      );
+      setTodoList([...revertedTodos]);
+    } finally {
+      setIsSaving(false);
+    }
   }
-  function handleUpdateTodo(editedTodo) {
+
+  async function completeTodo(id) {
```

Comment:
This is taking both optimistic and pessimistic strategies when only a optimistic strategy should be followed. So one `setTodoList` needs to be removed.